apostrophecms/apostrophe
# 1973 focal point

[PRO-1973](https://linear.app/apostrophecms/issue/PRO-1973/as-a-user-i-can-specify-a-focal-point-on-an-image-in-an-image-field)

## Summary

Handle focal point by moving it via drag&drop and clicking on the image.

Focal point coordinates are relative to the stencil (cropper rectangle), and are set as percentages.
For instance, clicking at the very center of the stencil, focal point coordinates will be `{ x: 50, y: 50 }`.

When clicking or moving it outside the stencil, coordinates will be sent as `{ x: null, y: null }`.

**Important note:** we cannot rely on the cropper coordinates that come from its API, because they do not reflect reality.
In other words, the cropper coordinates are updated when zooming on the image, but the focal position remains the same.
So taking the cropper coordinates provided by the API into account would give us wrong focal point relative (to the stencil rectangle) position.

## What are the specific steps to test this change?

- add an image and edit the relationship

👉  focal point should be at the very center of the stencil (default behaviour):

![image](https://user-images.githubusercontent.com/8301962/165100443-75586d58-9f7e-4d0e-bcf8-ae36a9adc859.png)

- click and/or drag the focal point to a part of the image you want (eg. this little plant):

![image](https://user-images.githubusercontent.com/8301962/165100950-d0729dc1-d4b3-40c1-bd1d-df8c0f55a5e2.png)

- update image relationship and save image widget editor
- go into preview mode and inspect image
- check that `object-position` values are coherent (eg. `object-position: 88% 88%;` in our example):

![image](https://user-images.githubusercontent.com/8301962/165101345-5a75ab7f-eb47-46e7-b251-b3af49b7f441.png)

- now add a `object-fit: cover` rule to the image element, and change its `width` and `height` in order to have a part of the image visible

👉  the part where the focal point has been set should be visible:

![image](https://user-images.githubusercontent.com/8301962/165102089-012daf9c-e571-4317-b796-ced75e37e1f5.png)

- edit the image relationship again

👉  the focal point should be placed at its last coordinates (not at the center anymore)

- place the focal point outside the stencil:

![image](https://user-images.githubusercontent.com/8301962/165102705-4b9dc186-4a4d-49ca-96bd-ed796a3438c0.png)

- add a `object-fit: cover` rule to the image element, and change its `width` and `height` in order to have a part of the image visible

👉  the image should not have a `object-position` property anymore:

![image](https://user-images.githubusercontent.com/8301962/165103212-65a48944-33cc-417f-a8fe-70dd4f3d688f.png)

- edit the image relationship again
- verify that moving the stencil updates the focal point `x` and `y` percentages (i.e. you do not have to click or move the focal point to take it into account)
- verify that resizing the screen places the focal point back at its position, relative to the stencil (i.e. the position percentages remain the same)
- check that moving and resizing the cropper stencil does not place the focal point at the clicked position (at least for 500ms, which is the debounce time) -- without that, UX is crap.

## What kind of change does this PR introduce?
*(Check at least one)*


## Make sure the PR fulfills these requirements:


If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**
